vermouth/molecule.py

```diff
@@ -33,8 +34,218 @@
+    correspondence is defined as the boolean result of the ``match`` method.
+
+    This is the base class for such predicate. It must be subclassed, and
+    subclasses must define a :meth:`match` method that takes a dictionary and
+    a potential key from that dictionary as arguments.
+
+    Parameters
+    ----------
+    value:
+        The per-instance value that serve as reference. How this value is
+        treated depends on the subclass.
+    """
+    def __init__(self, value):
+        self.value = value
+
+    def match(self, node, key):
+        """
+        Do the comparison with the reference value.
+
+        Notes
+        -----
+        This function **must** be defined by the subclasses. This docstring
+        describe the *expected* format of the method.
+
+        Parameters
+        ----------
+        node: dict
+            A dictionary of attributes in which to look up. This can be a
+            node dictionary of a molecule ``meta`` attribute.
+        key:
+            A potential key from the ``node`` dictionary.
+
+        Returns
+        -------
+        bool
+        """
+        
+        raise NotImplementedError
+
+    def __repr__(self):
+        return '<{} at {:x} value={}>'.format(self.__class__.__name__, id(self), self.value)
+
+
+class Choice(LinkPredicate):
+    """
+    Test if an attribute is defined and in a predefined list.
+
+    Parameters
+    ----------
+    value: list
+        The list of value in which to look for the attribute.
+    """
+    def match(self, node, key):
+        """
+        Apply the comparison.
+        """
+        return node.get(key) in self.value
+
+
+class NotDefinedOrNot(LinkPredicate):
+    """
+    Test if an attribute is not the reference value.
+
+    This test passes if the attribute is not defined, if it is set to ``None``,
+    or if its value is different from the reference.
+
+    Parameters
+    ----------
+    value:
+        The value the attribute is tested not to be.
+    """
+    def match(self, node, key):
+        """
+        Apply the comparison.
+        """
+        return key not in node or node[key] != self.value
+
+
+class LinkParameterEffector:
+    """
+    Rule to calculate an interaction parameter in a link.
+
+    This class allows to store dynamic parameters in link interactions. The
+    value of the parameter can be computed from the graph using the node keys
+    given when creating the instance.
+
+    An instance of this lass is first initialized with a list of node keys
+    from the import link in which it is defined. The instance is latter called
+    like a function, and takes as arguments a molecule and a match dictionary
+    linking the link nodes with the molecule ones. The format of the dictionary
+    is expected to be ``{link key: molecule key}``.
+
+    This is a base class; it needs to be subclassed. A subclass must define an
+    :meth:`_apply` method that takes a molecule and a list of node keys from
```

Comment:
I'm not convinced you need to define an underscore method. My preference would be either `apply` or `apply_`

Reply:
It was `apply` initially. But the method should not be used directly, which the underscore makes clear.